app/controllers/article_controller.rb

```diff
@@ -1,89 +1,66 @@
-    # Differences from previous implementation:
-    # 1. List of articles needs to be collection-specific
-    # 2. List should be displayed within the category treeview
-    # 3. Uncategorized articles should be listed below
-    if @collection.is_a?(DocumentSet)
-      @uncategorized_articles = @collection.articles.joins('LEFT JOIN articles_categories ac ON articles.id = ac.article_id').where('ac.category_id IS NULL').reorder(:title)
-    else
-      @uncategorized_articles = @collection.articles.where.not(:id => @collection.articles.joins(:categories).pluck(:id)).reorder(:title)
-    end
+    articles = @collection.articles.includes(:categories)
+    @categories = @collection.categories.includes(:articles)
 
-    # Create a 2D array of articles/subjects for each category
-    # where the articles are sorted alphabetically vertically
-    # instead of horizontally -- so, going down each column
-    @vertical_articles = {} # Holds articles for each category
+    @vertical_articles = {}
     @categories.each do |category|
-      articles = category.articles_list(@collection) # The articles for this category
-      @vertical_articles[category] = sort_vertically(articles)
+      current_articles = articles.where(categories: { id: category.id })
+      @vertical_articles[category] = sort_vertically(current_articles)
     end
 
-    # Do the same for uncategorized articles
-    if @uncategorized_articles.present?
-      @uncategorized_articles = sort_vertically(@uncategorized_articles)
-    end
+    @uncategorized_articles = sort_vertically(
+      articles.where(categories: { id: nil }).reorder(:title)
+    )
   end
 
   def delete
-    if @article.link_list.empty? && @article.target_article_links.empty?
-      if @article.created_by_id == current_user.id || current_user.like_owner?(@collection)
-        @article.destroy 
-        redirect_to collection_subjects_path(@collection.owner, @collection)
-      else
-        flash.alert = t('.only_subject_owner_can_delete')
-        redirect_to collection_article_show_path(@collection.owner, @collection, @article.id)
-      end
+    result = Article::Destroy.call(
+      article: @article,
+      user: current_user,
+      collection: @collection
+    )
+
+    if result.success?
+      redirect_to collection_subjects_path(@collection.owner, @collection)
     else
-      flash.alert = t('.must_remove_referring_links')
+      flash.alert = result.message
       redirect_to collection_article_show_path(@collection.owner, @collection, @article.id)
     end
   end
 
   def update
-    old_title = @article.title
-    gis_truncated = gis_truncated?(params[:article], GIS_DECIMAL_PRECISION)
-
-    @article.attributes = article_params
-    if params['save']
-      #process_source_for_article
-      if @article.save
-        if old_title != @article.title
-          rename_article(old_title, @article.title)
-        end
+    if params[:save]
+      result = Article::Update.call(
```

Comment:
I'd also like to know more about this syntax -- why use `Article::Update.call` instead of `@article.update`?

Reply:
The goal here is to simplify the logic in controllers by adopting a higher-level architecture.
The idea is, the controllers should only be concerned with routing, handling params, responses, etc. Anything CREATE/DESTROY (or what we can consider anything commiting to the db) is handed off to an interactor.

Interactors are meant to be simple, straightforward, and single purpose. This allows us developers, at a single glance on the files, to know what the application does.
Say in this example, for articles:
/interactors/article/
   - combine.rb
   - destroy.rb
   - update.rb
   
Now we know these 3 actions are available for article. This is a very simple example, but I can imagine we can have single purpose interactors like for isolating the logic on `transcription/update`, `transcription/mark_as_blank`, `transcription/set_needs_review`, which will help simplify and isolate what happens when these are done.

I also love this quote from the interactor-rails team on why use an interactor:

```
SPOILER ALERT: Your use case won't stay so simple.

In [our](http://collectiveidea.com/) experience, a simple task like authenticating a user will eventually take on multiple responsibilities:

    Welcoming back a user who hadn't logged in for a while
    Prompting a user to update his or her password
    Locking out a user in the case of too many failed attempts
    Sending the lock-out email notification

The list goes on, and as that list grows, so does your controller. This is how fat controllers are born.
``` 

I would say that, save_transcription route is one example of a fat controller, where there are more than 5 nested ifs, some with obscure params/conditions that will cause a redirect/render and some branches does CRUD operations while others don't. 

I think a higher architecture will help simplify and untangle things here, which will make updating/fixing bugs easier. 
Admittedly, the examples I gave are technically working (no bugs), and so this refactoring to simplify things are lower on the priority for me. But at the same time I think this is essential for higher coverage -> higher confidence on tests -> higher confidence in rails upgrade not breaking things


BTW:
Article::Update is a class, check interactors/article/update.rb I simply moved the logic there so that it is simpler to read for us.

Reply:
This makes sense for complex data manipulations like we see in the transcription controller and article controller (especially with things like `combine_duplicate`), though I'd prefer for us to not interactor-ize absolutely everything we touch.